### ISSUE TITLE
Pipeline propagation

### ISSUE THREAD

**Issue by caspervdw** [soft-matter/slicerator#6]
I ran into the following issue using the codes from the test:

```
v = Slicerator.from_list(list('abcdefghij'))
def _capitalize(letter):
    return letter.upper()
cap_b = pipeline(_capitalize)(v)
cap_b[1:][0]
```

Returns `'b'`. The processing function is not propagated.

**Comment by danielballan** [comment]
True! The pipeline function is, we see here, only lightly tested.

The solution is adding yet one more argument to `Slicerator.__init__`, which can be used to pass `self._proc_func` from parent to child.

**Comment by caspervdw** [comment]
Or we could nest the _get attributes, this would also allow for chained proc funcs. Each _get would call its ancestor _get.

**Comment by danielballan** [comment]
Yeah, this is tied into the question whether slicerator's should hold a reference to their ultimate ancestor or to their immediate parent, which you asked in a line comment elsewhere. I know I had a reason for doing it this way -- because I tried the other way first -- but I can't now recall what it was. I'll think about it. It might be that the reason is obsolete.

**Comment by caspervdw** [comment]
I see now, thanks that really clears some things up. The way it is currently implemented probably has less overhead, and I don't look forward to change the whole structure of `Slicerator`. We only need to cook up a way to have a chain of pipelines (and propagate it).

We could make `Slicerator._proc_func` a list of functions and indeed add it as yet another argument.

Maybe to shorten the list of arguments (a bit unorthodox) do the following inside `__getitem__`:

```
return Slicerator(self, indices, new_length)
```

And check at `__init__` if the first argument is instance of Slicerator, if so then pull ancestor, exposed attributes and process func from that.

**Comment by danielballan** [comment]
I'm coming around to thinking that holding a reference to the parent is cleaner, long-term, even though it involves a refactor that I'd rather avoid. The tests on the slicing logic are solid, so I feel on fairly safe ground reworking that.

The current implementation has less overhead, but unless users have slices ~100 deep, I doubt it makes a dent in performance compared to whatever they are doing with their objects (especially if those objects are images).

This is worth more thought...just sharing my current thought process.

**Comment by caspervdw** [comment]
I think that chained processing would definitely be simpler if the slicerator would be completely recursive (holding reference to its parent).

Then we could even pull the 2 applications apart into two classes: Slicerator is created by slicing, Pipeline is created by applying a pipeline function.

As they both will have the same getitem logic, we could shift the whole key -> indices logic to the Slicerator **init**. In this way, the Slicerator does not take an indices parameter but a slice/iterable/bool mask. Which would simplify the Slicerator construction API.

Just sharing my thoughts, I am open for other approaches.

**Comment by danielballan** [comment]
Before we put any work into refactoring, I'd like to understand better how PIMS' functionality maps onto the abstractions in dask. And that comes after merging #5, tagging a new 0.9.x, and tagging PIMS.

**Comment by caspervdw** [comment]
Good point, but still I would like to see this issue solved. I think forwarding proc_funcs would be an easy solution, I will try to implement that in #5

**Comment by danielballan** [comment]
Closed by #5
